### PR TITLE
GraphQL

### PR DESCRIPTION
**Pull request by trinitrot0luene**

### CODE REVIEW

**Comment by trinitrot0luene** [comment]
@Izumemori I'm a bit conflicted over exactly how to handle the nested nature of some of this data- for example Node -> MemberMetadata -> UserGroups is currently returning an array of usergroup IDs since Node -> UserGroups exists. Should the API always return the full-fat object or avoid returning duplicate data for nested queries like this?

**Comment by trinitrot0luene** [comment]
I suppose ```json
node {
    member_metadata { usergroups { id } }
}
```
would be a valid query, it'd make sense to leave it up to the consumer in that situation.

**Comment by Izumemori** [comment]
Generally should only be returned what is requested, with that example the json should be 
```json
{
  "node": {
     "member_metadata": {
       "usergroups": [
          {
            "id": 12121231313131
          },
          {
            "id": ...
          },
          ...
       ]
     }
  }
}
```

File: src/Scar.Api/Controllers/ApiController.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+        private readonly IDocumentExecuter _executer;
+
+        private readonly IDocumentWriter _writer;
+
+        private readonly ISchema _schema;
+
+        public ApiController(IDocumentExecuter executer, IDocumentWriter writer, ISchema schema)
+        {
+            _executer = executer;
```
Comment (Izumemori) [inline src/Scar.Api/Controllers/ApiController.cs]: I generally always prefix class members with 'this', if we want to change that we can

File: src/Scar.Api/Controllers/ApiController.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            _writer = writer;
+            _schema = schema;
+        }
+
+        [HttpPost]
+        public async Task<string> Query([FromBody] GraphQLQuery query)
+        {
+            var inputs = query.Variables.ToInputs();
+            var result = await _executer.ExecuteAsync(_ =>
+            {
```
Comment (Izumemori) [inline src/Scar.Api/Controllers/ApiController.cs]: this looks too much like discard syntax which is confusing

File: src/Scar.Api/Startup.cs [hunk 3]
```diff
@@ -65,7 +73,28 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
             }
 
             app.UseHttpsRedirection();
+            app.UseGraphiQl(path: "/graphql", apiPath: "/api");
             app.UseMvc();
         }
+
+        private void RegisterGraphQLServices(IServiceCollection services)
+        {
+            services.AddSingleton<IDocumentExecuter, DocumentExecuter>();
```
Comment (Izumemori) [inline src/Scar.Api/Startup.cs]: stack the AddSingleton calls

File: src/Scar.Api/Controllers/ApiController.cs (hunk 1 above)
Comment (trinitrot0luene) [reply src/Scar.Api/Controllers/ApiController.cs]: That's fine, done.

File: src/Scar.Api/Controllers/ApiController.cs (hunk 2 above)
Comment (trinitrot0luene) [reply src/Scar.Api/Controllers/ApiController.cs]: True, I was following the official documentation. Fixed!

File: src/Scar.Api/Startup.cs (hunk 3 above)
Comment (trinitrot0luene) [reply src/Scar.Api/Startup.cs]: Done

**Pull request merged**